pgmpy/utils/VariableStateNameDecorators.py

```diff
@@ -0,0 +1,119 @@
+"""
+The file contains various decorators to map variable state names
+to the internal representation of the node states.
+"""
+from itertools import product
+import six
+
+from pgmpy.extern import tabulate
+
+class stateNameInit():
+    """
+    Adds an optional state name dictionary as an attribute to the Factor class.
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from pgmpy.factors import Factor
+    >>> state_name_dict = {'x1':('on', 'off'), 'x2':('low', 'medium', 'high'), 
+                           'x3':('day', 'night')}
+    >>> phi = Factor(['x1', 'x2', 'x3'], [2, 2, 2], np.ones(8), state_name_dict)
+    >>> print(phi.state_names)
+    {'x1': ('on', 'off'), 'x2': ('low', 'medium', 'high'), 'x3': ('day', 'night')}
+
+    """
+    def __call__(self, f):
+        def init_wrapper(init_self, variables, cardinality, values, state_names=None):
+            init_self.state_names = state_names
+            f(init_self, variables, cardinality, values)
+        return init_wrapper
+
+
+class stateNameReduce():
+    """
+    Allows the reduce method to have state names in the input parameters.
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from pgmpy.factors import Factor
+    >>> state_name_dict = {'x1':('on', 'off'), 'x2':('low', 'medium', 'high'), 
+                           'x3':('day', 'night')}
+    >>> phi = Factor(['x1', 'x2', 'x3'], [2, 3, 2], np.ones(12), state_name_dict)
+    >>> print(phi.state_names)
+    {'x1': ('on', 'off'), 'x2': ('low', 'medium', 'high'), 'x3': ('day', 'night')}
+    >>> phi.reduce([('x1', 'on'), ('x2', 'low')])
+    >>> print(phi)
+    ╒═══════════╤═══════════╕
+    │ x3        │   phi(x3) │
+    ╞═══════════╪═══════════╡
+    │ x3(day)   │    1.0000 │
+    ├───────────┼───────────┤
+    │ x3(night) │    1.0000 │
+    ╘═══════════╧═══════════╛
```

Comment:
@khalibartan These are just temporary docs that I put for clarity. Besides,
the entire class will be reimplemented.
On 26 Feb 2016 14:28, "Utkarsh" <notifications@github.com> wrote:

> In pgmpy/utils/VariableStateNameDecorators.py
> <https://github.com/pgmpy/pgmpy/pull/648#discussion_r54218051>:
>
> > +    >>> import numpy as np
> > +    >>> from pgmpy.factors import Factor
> > +    >>> state_name_dict = {'x1':('on', 'off'), 'x2':('low', 'medium', 'high'),
> > +                           'x3':('day', 'night')}
> > +    >>> phi = Factor(['x1', 'x2', 'x3'], [2, 3, 2], np.ones(12), state_name_dict)
> > +    >>> print(phi.state_names)
> > +    {'x1': ('on', 'off'), 'x2': ('low', 'medium', 'high'), 'x3': ('day', 'night')}
> > +    >>> phi.reduce([('x1', 'on'), ('x2', 'low')])
> > +    >>> print(phi)
> > +    ╒═══════════╤═══════════╕
> > +    │ x3        │   phi(x3) │
> > +    ╞═══════════╪═══════════╡
> > +    │ x3(day)   │    1.0000 │
> > +    ├───────────┼───────────┤
> > +    │ x3(night) │    1.0000 │
> > +    ╘═══════════╧═══════════╛
>
> Use python 2 print output here, instead of python3
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pgmpy/pgmpy/pull/648/files#r54218051>.
>